### PR TITLE
account for an s3 folder when listing objects

### PR DESCRIPTION
**Pull request by briandowns**
#### Proposed Changes ####

Additional logic needed to be added to account for the use of folders within AWS S3 buckets in the "list snapshots" routine.

#### Verification ####

Run the command below and verify that what is listed in the ConfigMap is what's in the AWS folder. 

```sh
kubectl get cm -n kube-system k3s-etcd-snapshots -o yaml 
```

#### Linked Issues ####

https://github.com/rancher/rke2/issues/1551

#### User-Facing Change ####

```release-note

```

### CODE REVIEW

File: pkg/etcd/etcd.go [hunk 1]
```diff
@@ -904,7 +904,16 @@ func (e *ETCD) listSnapshots(ctx context.Context, snapshotDir string) ([]Snapsho
 			return nil, err
 		}
 
-		objects := e.s3.client.ListObjects(ctx, e.config.EtcdS3BucketName, minio.ListObjectsOptions{})
+		s3FolderInUse := e.config.EtcdS3Folder != ""
```
Comment (brandond) [inline pkg/etcd/etcd.go:907]: See below
```suggestion
```

File: pkg/etcd/etcd.go [hunk 2]
```diff
@@ -904,7 +904,16 @@ func (e *ETCD) listSnapshots(ctx context.Context, snapshotDir string) ([]Snapsho
...
+
+		var loo minio.ListObjectsOptions
+		if s3FolderInUse {
+			loo = minio.ListObjectsOptions{
+				Recursive: true,
+			}
```
Comment (brandond) [inline pkg/etcd/etcd.go:913]: Pass in the prefix, there's no point in listing the whole bucket from the root if you know everything's going to be under this folder prefix
```suggestion
			loo = minio.ListObjectsOptions{
				Prefix: e.config.EtcdS3Folder,
				Recursive: true,
			}
```

File: pkg/etcd/etcd.go [hunk 3]
```diff
@@ -914,6 +923,12 @@ func (e *ETCD) listSnapshots(ctx context.Context, snapshotDir string) ([]Snapsho
 				continue
 			}
 
+			if s3FolderInUse {
+				if !strings.Contains(obj.Key, e.config.EtcdS3Folder) {
+					continue
+				}
+			}
```
Comment (brandond) [inline pkg/etcd/etcd.go:930]: Unnecessary if you pass in the prefix

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/etcd/etcd.go (hunk 2 above)
Comment (briandowns) [reply pkg/etcd/etcd.go:913]: Sadly, none of the prefix stuff is working in this library or is in a very non-obvious way. The prefix was the preferred route.

Comment (brandond) [reply pkg/etcd/etcd.go:913]: Hmm, listing with a prefix is pretty important to the S3 API. Do you perhaps need to strip the trailing slash or something? Client-side filtering by checking the key is going to be terrible; buckets can contain millions of objects.

Comment (brandond) [reply pkg/etcd/etcd.go:913]: https://docs.min.io/docs/golang-client-api-reference#:~:text=prefix%3A%20%22myprefix%22

Comment (brandond) [reply pkg/etcd/etcd.go:913]: if the object is bucket/foo/bar.txt, the prefix should be specified as "foo", not "/foo" or "foo/".

Comment (briandowns) [reply pkg/etcd/etcd.go]: Was the missing recursive field. ugh

**Review by brandond — APPROVED** [review]

**Comment by brandond** [comment]
Does this currently find everything properly if there's not a prefix specified? I guess it just non-recursively lists everything in the top-level directory (no prefix)?

**Comment by briandowns** [comment]
> Does this currently find everything properly if there's not a prefix specified? I guess it just non-recursively lists everything in the top-level directory (no prefix)?

Correct. Only looks at the top level (bucket level).

**Review by Oats87 — APPROVED** [review]

**Pull request merged**